### PR TITLE
Serialization issues on OneSdk

### PR DESCRIPTION
**Pull request by gcorvera**
Fixes issues with Serialization. Also opens Key in Things to enable external apps setting it.

### CODE REVIEW

File: Microsoft.HealthVault/Thing/ThingBase.cs [hunk 1]
```diff
@@ -835,7 +835,9 @@ public string Serialize()
             XDocument newDocument = SDKHelper.SafeLoadXml(GetItemXml());
 
             string dataName = "thing/data-xml";
-            fetchedDocument.Element(dataName).ReplaceWith(newDocument.Element(dataName));
```
Comment (suryaren) [inline Microsoft.HealthVault/Thing/ThingBase.cs]: >Element [](start = 28, length = 7)

what is cause for the change ? looks like if there is a problem then updates wont happen .. but that's not the case .. we are able to update ..

File: Microsoft.HealthVault/Thing/ThingKey.cs [hunk 2]
```diff
@@ -87,7 +87,10 @@ public ThingKey(Guid id)
         /// issued when the item is created.
         /// </value>
         ///
-        public Guid Id => _thingId;
+        public Guid Id {
```
Comment (suryaren) [inline Microsoft.HealthVault/Thing/ThingKey.cs]: >{ [](start = 23, length = 1)

nit: next line ..

File: Microsoft.HealthVault/Thing/ThingKey.cs [hunk 3]
```diff
@@ -106,7 +109,17 @@ public ThingKey(Guid id)
         /// is always equal to the <see cref="Id"/> of that item.
         /// </remarks>
         ///
-        public Guid VersionStamp => _versionStamp;
+        public Guid VersionStamp
+        {
+            get { return _versionStamp; }
+            set
+            {
+                if(value != Guid.Empty)
```
Comment (suryaren) [inline Microsoft.HealthVault/Thing/ThingKey.cs]: >if(value != Guid.Empty) [](start = 16, length = 23)

Guiddefault value is empty .. so wondering if a check is required ..

File: Microsoft.HealthVault.UnitTest/ItemTypes/WeightTests.cs [hunk 4]
```diff
@@ -31,5 +31,16 @@ public void WhenSerializesXml_ThenCorrectSerializationReturned()
             Assert.AreEqual("kg", weight.Value.DisplayValue.Units);
             Assert.AreEqual("kg", weight.Value.DisplayValue.UnitsCode);
         }
+
+        [TestMethod]
+        public void WhenDeserializesXml_ThenCorrectStringReturned()
```
Comment (suryaren) [inline Microsoft.HealthVault.UnitTest/ItemTypes/WeightTests.cs]: >Deserializes [](start = 24, length = 12)

serialized is what you are checking for ?

File: Microsoft.HealthVault/Thing/ThingBase.cs (hunk 1 above)
Comment (gcorvera) [inline Microsoft.HealthVault/Thing/ThingBase.cs]: Element("path") doesn't support "/" characters, in other words you are selecting an specific Element by name. When this was written, using thing/data-xml was meant to work as a xpath. xpathselectelement is more appropriate for this. 

Could you be more specific on what you mean by: " looks like if there is a problem then updates wont happen .. but that's not the case .. we are able to update .." I am not sure I understood what you mean.  

---
In reply to: [121809660](https://github.com/Microsoft/healthvault-dotnetstandard-sdk/pull/57#discussion_r121809660) [](ancestors = 121809660)

File: Microsoft.HealthVault/Thing/ThingKey.cs (hunk 3 above)
Comment (gcorvera) [inline Microsoft.HealthVault/Thing/ThingKey.cs]: I think we should just not assign it. There is no reason to fail at this point. 

---
In reply to: [121810435](https://github.com/Microsoft/healthvault-dotnetstandard-sdk/pull/57#discussion_r121810435) [](ancestors = 121810435)

File: Microsoft.HealthVault/Thing/ThingBase.cs (hunk 1 above)
Comment (suryaren) [inline Microsoft.HealthVault/Thing/ThingBase.cs]: nvm, I thought that this method is being used for the update operation to platform.

File: Microsoft.HealthVault/Thing/ThingKey.cs (hunk 3 above)
Comment (suryaren) [inline Microsoft.HealthVault/Thing/ThingKey.cs]: Not sure if I follow you.. where is the failure here ..

**Comment by HATIMS** [comment]
:shipit:

**Pull request merged**